### PR TITLE
feat(TextArea, TextInput): use aria-describedby for error and note

### PR DESCRIPTION
**Pull request by Kyzyl-ool**
closes https://github.com/gravity-ui/uikit/issues/860

### CODE REVIEW

**Comment by korvin89** [comment]
Could U add this changes to TextArea too?

**Comment by Kyzyl-ool** [comment]
> Could U add this changes to TextArea too?

Ready

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: src/components/controls/TextInput/TextInputControl.tsx [hunk 1]
```diff
@@ -30,8 +31,24 @@ export function TextInputControl(props: Props) {
...
+            if (note) {
+                result.push(getControlNoteId(id));
+            }
+
+            if (error) {
+                result.push(getControlErrorTextId(id));
+            }
+        }
+        return result.join(' ');
+    }, [error, note, id]);
```
[the remark this answers is not in the document]
Comment (Kyzyl-ool) [reply src/components/controls/TextInput/TextInputControl.tsx]: You are right, just fixed it. Thank you!

File: src/components/controls/TextArea/TextArea.tsx [hunk 2]
```diff
@@ -73,6 +75,13 @@ export const TextArea = React.forwardRef<HTMLSpanElement, TextAreaProps>(functio
     const isClearControlVisible = Boolean(hasClear && !disabled && inputValue);
     const id = originalId || innerId;
 
+    const ariaDescribedBy = useDescribedBy({
+        ariaDescribedBy: controlProps?.['aria-describedby'],
+        note,
+        error,
+        controlId: id,
+    });
```
Comment (ValeraS) [inline src/components/controls/TextArea/TextArea.tsx:83]: Why should ids of error message and note depend on controlId?

```suggestion
    const errorMessageId = useUniqId();
    const noteId = useUniqId();
    const ariaDescribedBy = [
        controlProps?.['aria-describedby'],
        note ? noteId : undefined,
        isErrorMsgVisible ? errorMessageId : undefined
    ].filter(Boolean).join(' ');
```

File: src/components/controls/TextArea/TextArea.tsx [hunk 3]
```diff
@@ -159,12 +172,12 @@ export const TextArea = React.forwardRef<HTMLSpanElement, TextAreaProps>(functio
...
-                    {isErrorMsgVisible && <div className={b('error')}>{error}</div>}
-                    {note && <div className={b('note')}>{note}</div>}
-                </div>
-            )}
+            <OuterAdditionalContent
+                note={note}
+                error={error}
+                isVisible={isErrorMsgVisible}
+                controlId={id}
+            />
```
Comment (ValeraS) [inline src/components/controls/TextArea/TextArea.tsx:180]:
```suggestion
            <OuterAdditionalContent
                note={note}
                noteId={noteId}
                errorMesage={isErrorMsgVisible ? error : undefined}
                errorMessageId={errorMessageId}
            />
```

File: src/components/controls/TextInput/__tests__/TextInput.input.test.tsx [hunk 4]
```diff
@@ -125,6 +126,29 @@ describe('TextInput input', () => {
...
+                // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
+                const errorTextElement = container.querySelector(
+                    `#${getControlErrorTextId(inputId)}`,
+                );
+                // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
+                const noteTextElement = container.querySelector(`#${getControlNoteId(inputId)}`);
+
+                expect(input.getAttribute('aria-describedby')).toBe(
+                    `${getControlNoteId(inputId)} ${getControlErrorTextId(inputId)}`,
+                );
```
Comment (ValeraS) [inline src/components/controls/TextInput/__tests__/TextInput.input.test.tsx:148]:
```suggestion
                const [noteId = '', errorMessageId = ''] = (
                    input.getAttribute('aria-describedby') ?? ''
                ).split(/\s+/);
                expect(noteId).not.toBe('');
                expect(errorMessageId).not.toBe('');

                const errorTextElement = queryHelpers.queryByAttribute(
                    'id',
                    container,
                    errorMessageId,
                );
                const noteTextElement = queryHelpers.queryByAttribute('id', container, noteId);
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: src/components/controls/TextArea/TextArea.tsx [hunk 5]
```diff
@@ -90,7 +99,11 @@ export const TextArea = React.forwardRef<HTMLSpanElement, TextAreaProps>(functio
             }
         },
         autoComplete: prepareAutoComplete(autoComplete),
-        controlProps,
+        controlProps: {
+            ...controlProps,
+            'aria-describedby': ariaDescribedBy,
+            'aria-invalid': controlProps?.['aria-invalid'] ?? Boolean(error),
```
[the remark this answers is not in the document]
Comment (Kyzyl-ool) [reply src/components/controls/TextArea/TextArea.tsx:105]: Agree

File: src/components/controls/TextArea/TextArea.tsx (hunk 3 above)
Comment (Kyzyl-ool) [reply src/components/controls/TextArea/TextArea.tsx:180]: 👍

File: src/components/controls/TextInput/__tests__/TextInput.input.test.tsx (hunk 4 above)
Comment (Kyzyl-ool) [reply src/components/controls/TextInput/__tests__/TextInput.input.test.tsx:148]: Useful, thank you

**Review by ValeraS — APPROVED** [review]

File: src/components/controls/TextArea/TextArea.scss [hunk 6]
```diff
@@ -46,31 +46,6 @@ $block: '.#{variables.$ns-new}text-area';
         }
     }
 
-    &__outer-additional-content {
```
Comment (amje) [inline src/components/controls/TextArea/TextArea.scss:49]: Shouldn't we also remove styles from 'TextInput.scss'?

File: src/components/controls/TextArea/TextArea.tsx [hunk 7]
```diff
@@ -104,7 +110,11 @@ export const TextArea = React.forwardRef<HTMLSpanElement, TextAreaProps>(functio
             }
         },
         autoComplete: prepareAutoComplete(autoComplete),
-        controlProps,
+        controlProps: {
+            ...controlProps,
+            'aria-describedby': ariaDescribedBy || undefined,
+            'aria-invalid': Boolean(error) || undefined,
```
Comment (amje) [inline src/components/controls/TextArea/TextArea.tsx:116]: We should check `validationState` instead

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/controls/TextArea/TextArea.scss (hunk 6 above)
Comment (Kyzyl-ool) [reply src/components/controls/TextArea/TextArea.scss:49]: Of course. Missed

**Review by amje — APPROVED** [review]

**Pull request merged**